module/os/linux/zfs/vdev_file.c

```diff
@@ -248,18 +252,14 @@ vdev_file_io_start(zio_t *zio)
 		zio_execute(zio);
 		return;
 	} else if (zio->io_type == ZIO_TYPE_TRIM) {
-		struct flock flck;
+		int mode = 0;
 
 		ASSERT3U(zio->io_size, !=, 0);
-		bzero(&flck, sizeof (flck));
-		flck.l_type = F_FREESP;
-		flck.l_start = zio->io_offset;
-		flck.l_len = zio->io_size;
-		flck.l_whence = SEEK_SET;
-
-		zio->io_error = VOP_SPACE(vf->vf_vnode, F_FREESP, &flck,
-		    0, 0, kcred, NULL);
-
+#ifdef FALLOC_FL_PUNCH_HOLE
+		mode = FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE;
+#endif
```

Comment:
While you're here can you remove the `FALLOC_FL_PUNCH_HOLE` and `FALLOC_FL_KEEP_SIZE` guards.  This functionality was introduced in the 2.6.38 kernels and is now expected to exist in all supported Linux kernels.